### DISCUSSION TITLE
Add Support for Pydantic models in multipart/form data requests.

### DISCUSSION THREAD

**Discussion by pythonweb2** [fastapi/fastapi#8406] [Questions] [answered]
### Example Code

```python
from typing import List, Type, Union, Any, Dict, Tuple, TypedDict

from fastapi import Depends, FastAPI
from fastapi import Form, UploadFile, File
from fastapi.openapi.utils import get_openapi
from pydantic import BaseModel
from pydantic import ValidationError
from starlette.responses import JSONResponse
from starlette.status import HTTP_422_UNPROCESSABLE_ENTITY

Loc = Tuple[Union[int, str], ...]


class _ErrorDictRequired(TypedDict):
    loc: Loc
    msg: str
    type: str


class ErrorDict(_ErrorDictRequired, total=False):
    ctx: Dict[str, Any]


app = FastAPI()


class MultipleValidationErrors(Exception):
    def __init__(self, errors: List[ErrorDict]):
        super().__init__()
        self.errors = errors


class TestSchema(BaseModel):
    foo: str
    bar: int


def validate_json(model: Type[BaseModel]):
    def wrapper(body: str = Form(...)):
        try:
            return model.parse_raw(body)
        except ValidationError as exc:
            errors = exc.errors()
            for error in errors:
                error["loc"] = tuple(["body"] + list(error["loc"]))
            raise MultipleValidationErrors(errors)

    return wrapper


@app.post("/test")
def endpoint(file: UploadFile = File(...), body: TestSchema = Depends(validate_json(TestSchema))):
    print(file)
    print(body)
    return body


@app.exception_handler(MultipleValidationErrors)
def exception_handler(_request, exc: MultipleValidationErrors):
    return JSONResponse(status_code=HTTP_422_UNPROCESSABLE_ENTITY, content={"detail": exc.errors})


def custom_openapi():
    if app.openapi_schema:
        return app.openapi_schema
    openapi_schema = get_openapi(title="Test App", version="v1", routes=app.routes)

    # Provide Schema for JSON body inside multipart/form type
    openapi_schema["components"]["schemas"]["Body_endpoint_test_post"]["properties"]["body"] = {
        "title": "Body",
        "type": "object",
        "$ref": "#/components/schemas/TestSchema",
    }
    openapi_schema["components"]["schemas"]["TestSchema"] = TestSchema.schema()

    app.openapi_schema = openapi_schema
    return app.openapi_schema


app.openapi = custom_openapi
```

### Description

The documentation at https://fastapi.tiangolo.com/tutorial/request-forms-and-files states:

> You can declare multiple File and Form parameters in a path operation, but you can't also declare Body fields that you expect to receive as JSON, as the request will have the body encoded using multipart/form-data instead of application/json.

> This is not a limitation of FastAPI, it's part of the HTTP protocol.

This is true, but there is a workaround if you want to send data along with a file in a multipart/form request. You can send a string of JSON data along with the request, and both swagger-ui and redoc support this sort of openapi definition. For example, in the example code above, this is what I see in the documentation pages.

![image](https://user-images.githubusercontent.com/32141163/199875889-2c564052-2421-4068-97f8-c86c0f590bc3.png)
![image](https://user-images.githubusercontent.com/32141163/199875954-2c08d764-8442-4714-b0d8-9ef2a4fba1b4.png)

In the example, I had to manually change the openapi schema file in order to get this to work, but it is certainly supported at this time.

### Wanted Solution

I was hoping that FastAPI could natively support this approach to sending a JSON string that could be automatically documented and validated, as opposed to needing to go through these workarounds that I implemented in my example code.

As you can see in my example code below, this would offer a concise solution for anyone who want to accomplish this in their endpoints.

### Wanted Code

```python
from fastapi import FastAPI
from fastapi import File
from fastapi import FormObject  # New way to declare that you want an openapi object in the form
from fastapi import UploadFile
from pydantic import BaseModel

app = FastAPI()


class TestSchema(BaseModel):
    foo: str
    bar: int

@app.post("/test")
def endpoint(file: UploadFile = File(...), body: TestSchema = FormObject(...)):
    print(file)
    print(body)
    return body
```

### Operating System

Linux, Windows

### FastAPI Version

0.74.1

### Python Version

Python 3.9.12

**Comment by rajasimon** [comment]
Literally facing the same issue where I need to capture the incoming POST method body that has this `application/x-www-form-urlencoded` header. 

So far it can only capture the string that I mentioned in the parameter. `async def index(command: str = Form()):` this will print command as it is. 

Now my question is how do I capture all of the incoming data into the class instance? 


```
# This class instance never works and it throws `INFO:     54.157.105.70:0 - "POST / HTTP/1.1" 422 Unprocessable Entity`
class Command(BaseModel):
    token: str


# b'token=thisistoken&team_id=TEAMID213&team_domain=bwork&channel_id=IUHFF32HF&channel_name


@app.post("/")
async def index(command: str = Form(...)):
    # body = await request.body()
    print(command)
    import pdb

    pdb.set_trace()
```

**Comment by iudeen** [comment]
You can do this using both `dataclasses`
```python
from dataclasses import dataclass
from fastapi import FastAPI, Form, Depends, UploadFile
from starlette.responses import HTMLResponse

app = FastAPI()



@dataclass
class SimpleModel:
    no: int = Form(...)
    nm: str = Form(...)
    f: UploadFile = Form(...)


@app.post("/form")
def form_post(form_data: SimpleModel = Depends()):
    return form_data
```

or using Pydantic `BaseModel`

```python
from fastapi import FastAPI, Form, Depends, UploadFile
from pydantic import BaseModel

app = FastAPI()


class SimpleModel(BaseModel):
    no: int = Form(...)
    nm: str = Form(...)
    f: UploadFile = Form(...)


@app.post("/form")
async def form_post(form_data: SimpleModel = Depends()):
    fp = await form_data.f.read()
    print(fp)
```

**Reply by maradeben** [reply]
I have been on this for hours. BaseModel didn't work, but dataclasses did the magic

**Reply by eegli** [reply]
@maradeben 
FINALLY! This is the answer. I'd love to know why it does not work for `BaseModel`s, though.

**Reply by marcin-lulek-cint** [reply]
Any idea why this doesn't work? @tiangolo Is this a bug that it works only for `dataclasses`?

**Comment by rajasimon** [comment]
I'm still getting the `422 Unprocessable Entity`

**Comment by pythonweb2** [comment]
@iudeen your example is not the same, unfortunately.

The OpenAPI document generated is different, and this would not support recursive Pydantic models, nor would it support Pydantic custom validators.

The change that I am suggesting would allow a user to drop in a complex Pydantic model, and have it behave the exact same way as it would in a request where it was used as a request body.

**Reply by sanzoghenzo** [reply]
HTML forms can't be nested, thus direct support of recursive Pydantic models isn't possible.
The solution for a simple pydantic model in [this discussion](https://github.com/tiangolo/fastapi/discussions/7268), and a library that supports pydantic v1 can be found [here](https://github.com/dotX12/pyfa-converter).

**Reply by pythonweb2** [reply]
@sanzoghenzo that isn't right, it is possible, and I showed how to support it in my example. Basically just pass a string of JSON in a form string type field and use pydantic to validate it.

**Reply by sanzoghenzo** [reply]
Oh sorry, I was thinking more from an html `<form>` perspective... of course you can put a json inside a multipart form data.

I see now that this would come in handy from time to time (for now I would be good with a built-in way to turn "flat" models to form fields).

**Comment by ornakash** [comment]
Does anyone have a solution for this?

**Reply by jayashankarjayan** [reply]
Not yet, but I am giving it a try

**Reply by pythonweb2** [reply]
In my question I have a workaround, but there is no native support right now.

**Reply by kubasaw** [reply]
I think there **was** quite powerful solution, based on pydantic type `Json`

```python
from pydantic import BaseModel, Json
from fastapi import FastAPI, Form, UploadFile
from uvicorn import run


class TestModel(BaseModel):
    a: int
    b: str


app = FastAPI()


@app.post("/")
def endpoint(model: Json[TestModel] = Form(), file: UploadFile = None):
    return model


if __name__ == "__main__":
    run("main:app", reload=True)

```

You can test it with:
```python
import httpx
import json

response = httpx.post('http://localhost:8000', data={
    'model':json.dumps({'a':1,'b':'text'})
})

print(response.json()) #returns {'a':1,'b':'text'}
```

It works only for **pydantic@v1**: generated OpenAPI schema is also correct.

After upgrading to **pydantic@v2**, FastAPI returns:
```json
{
   "detail":[
      {
         "type":"model_attributes_type",
         "loc":[
            "body",
            "model"
         ],
         "msg":"Input should be a valid dictionary or object to extract fields from",
         "input":"{\"a\": 1, \"b\": \"text\"}",
         "url":"https://errors.pydantic.dev/2.0.3/v/model_attributes_type"
      }
   ]
}
```
In my personal case, this does not allow any upgrade to the pydantic@2 version. Issue is caused by FastAPI glue-code for pydantic. `Model` is unpacked from `Json[]`, so pydantic model is created as `Model` not `Json[Model]` which is able to load JSON from string/bytes (presented here: https://github.com/pydantic/pydantic/issues/6663)

The same error message is emitted when you use `Annotation` syntax with **pydantic@2**
```python
from typing import Annotated
from pydantic import BaseModel, Json
from fastapi import FastAPI, Form, UploadFile
from uvicorn import run


class TestModel(BaseModel):
    a: int
    b: str


app = FastAPI()


@app.post("/")
def endpoint(model: Annotated[Json[TestModel], Form], file: UploadFile = None):
    return model
```

When downgrade to **pydantic@1**, FastAPI not start returning
`AssertionError: Param: model can only be a request body, using Body()`

Same issue is described here: https://github.com/tiangolo/fastapi/discussions/9305

I think is huge regression, should be converted to Issue and fixed.

**Reply by wolegechu** [reply]
Try to add `headers = {'Content-Type': 'application/json'}` to `post` call.

**Comment by Matthieu-LAURENT39** [comment]
Still no workaround for pydantic v2?

**Reply by vgonisanz** [reply]
There is not a nice hack yet.  But there is a way to parse the elements, by adding "()" to the previous example in the Form, it is possible to parse the dictionary automatically.

- Pros: It works, it removes extra dictionary elements and asks for the missing elements.
- Cons: The Swagger UI asks for a string, and not the expected fields. It is necessary to enter them by hand.

using this example:

```python
from typing import Annotated
from pydantic import BaseModel, Json
from fastapi import FastAPI, Form, UploadFile
from uvicorn import run


class TestModel(BaseModel):
    a: int
    b: str


app = FastAPI()


@app.post("/")
def endpoint(model: Annotated[Json[TestModel], Form()], file: UploadFile = None):
    print(f"{model.a}, {model.b}")
    return model
```

And sending from swagger any file, and:

- The model string: `{ "a": 0,   "b": "string", "c": "string" }`, the result is `{ "a": 0,   "b": "string"}`, ignoring any additional form field.
- The model string: `{"a": 0}` fails because b is missing.

Same result using curl command:

```
curl -X 'POST' \                                           
  'http://127.0.0.1:8000/' \
  -H 'accept: application/json' \
  -H 'Content-Type: multipart/form-data' \
  -F 'model={   "a": 0,   "b": "string", "c": "string"}' \
  -F 'file=@app.py;type=audio/x-wav'
  ```

**Reply by btx** [reply]
I was trying to make this work. After some experimentation and many combinations I've tried, I've found another solution. It's presumably a bit hacky as well, but close enough, I think:

```python
import enum
from typing import Annotated, Any

from fastapi import FastAPI, UploadFile, File
from fastapi.params import Form
from pydantic import BaseModel, NonNegativeFloat, Json

app = FastAPI()


class Language(enum.StrEnum):
    cs = enum.auto()
    en = enum.auto()


class UploadParameters(BaseModel):
    language: Language
    range_from: NonNegativeFloat = 0
    speedup: bool = False


class JsonForm(Json[Any], Form):  # type: ignore[misc]
    ...


@app.post(
    "/upload",
)
async def upload(
    file: Annotated[UploadFile, File()],
    params: Annotated[UploadParameters, JsonForm()],
) -> dict[str, Any]:
    return {
        "file": {
            "name": file.filename,
        },
        "params": params.model_dump(),
    }

```

The 'magic' is in `JsonForm` which inherits from `pydantic.Json` and `fastapi.params.Form`. I've also parametrized `Json` with `Any` and it was necessary to ignore an error to keep mypy silent.

Input field `params` is correctly validated and openapi.json seems correct. It looks nice in Swagger/Stopligt and a correct example value is provided for trying it out.

Now, my question is if this is really a way to go. Am I using public and stable API? Will this be supported and documented? If not, what's exactly wrong with the solution?

What do you think?

Dependencies:
```
[tool.poetry.dependencies]
python = "^3.11"
fastapi = "^0.103.1"
uvicorn = "^0.23.2"
python-multipart = "^0.0.6"
black = "^23.9.1"
mypy = "^1.5.1"
```

Thank you

**Reply by jqueguiner** [reply]
an even better solution to have the default values:
`params: Annotated[UploadParameters, JsonForm()] = json.dumps(vars(UploadParameters()))`

**Reply by kheyer** [reply]
is this tied to a feature in python 3.11? I'm trying it with python 3.9 and getting the following error on `JsonForm`

```
TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
```

**Comment by kheyer** [comment]
I was able to get this working using the following:

```python
class UploadData(BaseModel):
    file: UploadFile = Form(...)
    max_file_size: Optional[int] = Form(None)
    created_by: Optional[str] = Form(None)
    filename: Optional[str] = Form(None)

@router.post('/upload')
async def upload(form_data: UploadData = Depends()):
    ...
```

One thing I found was the order of items in the pydantic schema matters. 

This works:

```python
class UploadData(BaseModel):
    file: UploadFile = Form(...)
    max_file_size: Optional[int] = Form(None)
    created_by: Optional[str] = Form(None)
    filename: Optional[str] = Form(None)
```

This returns a 422 error:

```python
class UploadData(BaseModel):
    max_file_size: Optional[int] = Form(None)
    created_by: Optional[str] = Form(None)
    filename: Optional[str] = Form(None)
    file: UploadFile = Form(...)
```

**Reply by edwinmarinn** [reply]
With this approach, `max_file_size`, `created_by` and `filename` are query parameters but ideally these data should be sent in the body

**Reply by kheyer** [reply]
Is there a way to do that? Ideally I'd like something like 

```python
class UploadParameters(BaseModel):
    max_file_size: Optional[int]
    created_by: Optional[str]
    filename: Optional[str] 

@router.post('/upload')
async def upload(file: UploadFile, upload_params: UploadParameters):
    ...
```

But my understanding of the issue is this is not possible

**Reply by alvarolopez** [reply]
> With this approach, `max_file_size`, `created_by` and `filename` are query parameters but ideally these data should be sent in the body

If instead of using `pydantic.BaseModel` you use a dataclass, the `max_file_size`, `created_by` and `filename` are in the body. I have not managed to get this working with Pydantic V2.

**Comment by osangu** [comment]
I have same error when using pydantic v2.
And i found the code which raising the exception.

```python
# fastapi._combat.py

def validate(
    self,
    value: Any,
    values: Dict[str, Any] = {},  # noqa: B006
    *,
    loc: Tuple[Union[int, str], ...] = (),
) -> Tuple[Any, Union[List[Dict[str, Any]], None]]:
    try:
        return (
            self._type_adapter.validate_python(value, from_attributes=True),
            None,
         )

    except ValidationError as exc:
        return None, _regenerate_error_with_loc(
            errors=exc.errors(), loc_prefix=loc
        )
```

I think this error cause by type_adapter in `fastapi._compat.py`, or may be wrong design about TypeAdapter class.

By following bellow dependencies you can also find the code.

`get_route_handler(routing.py) -> get_request_handler(routing.py) -> solve_dependencies(utils.py) -> request_body_to_args(utils.py) -> validate(_compat.py)`

P. S
At _compat.py line 46, you can find `if` sentence seperating pydantic v2

**Comment by DragonGamees** [comment]
You can add to your BaseModel
```py
@model_validator(mode='before')
@classmethod
def validate_to_json(cls, value: Any) -> Any:
    print(value)
    if isinstance(value, str):
        return cls(**json.loads(value))
    return value
```
Example


```py
import json
from pydantic import BaseModel, model_validator



class TestSchema(BaseModel):
    foo: str
    bar: int

    @model_validator(mode='before')
    @classmethod
    def validate_to_json(cls, value: Any) -> Any:
        print(value)
        if isinstance(value, str):
            return cls(**json.loads(value))
        return value

@app.post("/test")
def endpoint(file: UploadFile = File(...), body: TestSchema = Form(...)):
    print(file)
    print(body)
    return body

```

**Reply by jan-zmeskal** [reply]
I faced the same issue recently and adapted your solution. This version achieves the same using only pydantic capabilities and is inspired by this [doc section](https://docs.pydantic.dev/latest/concepts/json/#partial-json-parsing). Notice that I don't return a model instance from the validator. I don't know pydantic internals all that well and I suspect that conversion to instance might lead to duplicate validation in some circumstances. Furthermore, the example from [pydantic docs](https://docs.pydantic.dev/latest/concepts/validators/#field-validators) returns primitive structure from the validator, so I decided to stick with that approach.

```python
from pydantic import BaseModel, model_validator
from pydantic_core import from_json


class TestSchema(BaseModel):
    foo: str
    bar: int

    @model_validator(mode="before")
    @classmethod
    def convert_json_string_into_dictionary(cls, data: Any) -> Any:
        if isinstance(data, str):
                return from_json(data)
        return data
```

**Reply by MattdaVill** [reply]
This works but holy guacamole I don't wanna push this to production.

**Reply by jan-zmeskal** [reply]
While I'd also love this to be acknowledged and tackled officially, what makes this approach unsuitable for production? I have no problem with it and auto-generated docs (stoplight specifically, didn't try swagger) render it in-depth.

**Reply by MattdaVill** [reply]
Because I am specifically changing the behaviour of pydantic for a single BaseModel. My future me (and anybody else working on this codebase) would hate me for it.

**Reply by MattdaVill** [reply]
Also I am still having several problems even with this solution. I might be wrong but if now try to send a post without a file (in my case is optional) pydantic goes bersek. I could only make it work by including and empty files field (I am using python 3.12 with httpx as lib).

**Reply by jan-zmeskal** [reply]
I'll try to show how we use it in our codebase. Since it's not open-source, I had to adapt it a bit before publishing. Please keep that in mind, it might not work exactly as-is.

Firstly, we define a base model using the trick I posted above:
```python
class JsonStringModel(BaseModel):
    """`BaseModel` subclass used to validate objects passed to API as JSON strings.

    The primary use case is treating one of the form fields as JSON payload. We need to do that
    in case of endpoints that accept both file and rich set of input parameters. We cannot receive
    both file and JSON body because they use conflicting Content-Type header. It is
    "multipart/form-data" for the former and "application/json" for the latter. If we require
    a client to send a file, we have to use the data sent in form for input parameters.

    Theoretically we could use several form fields to gather all the required input parameters.
    However, some of them are nested in their nature and forms don't support that. That's why we
    consider it better to just use one form field and process its content as JSON payload.

    If a form field model inherits from this class, we will get API documentation and input
    validation - just the way we get it for JSON body defined with `BaseModel`.
    """
    @model_validator(mode="before")
    @classmethod
    def convert_json_string_into_dictionary(cls, data: Any) -> Any:
        if isinstance(data, str):
            try:
                return from_json(data)
            except ValueError as e:
                raise ValueError("Input should be a valid JSON string") from e
        return data
```

Then, we define a model of data that we expect from user. It inherits from `JsonStringModel` instead of `BaseModel`, making it clear that there might be something on top of default pydantic behavior. 

```python
class AdvancedEndpointConfig(JsonStringModel):
    list_of_strings: list[str]
    list_of_other_models: list[OtherComplexModel]
```

Finally, we include it in an endpoint somewhat like this:
```python
@router.post("/api/endpoint")
async def handler(file: UploadFile, config: AdvancedEndpointConfig = Form(...)):
    ...
```

Like this, `file` and `config` are independent of each other.

**Reply by MattdaVill** [reply]
I'll test it, thanks for sharing!

**Reply by TomFaulkner** [reply]
I tried this from the Swagger, and it works. But, it's still sending a json string rather than how a browser would send a form. In Firefox it sends this:

`{"a":0,"b":"string"}`

**Reply by bl1nkker** [reply]
the savior of my time

this was the answer for me

**Reply by Brikas** [reply]
With this method I am having documentation with Swagger, where for the form data, it says "object" and does not document the model (what kind of key should the object have, etc.)

**Comment by etiennebonnafoux** [comment]
I have the same issue. Here is my code

```python
@router.post("/add_project_user")
def add_project_user(
    status: str,
    name: str,
    email: str,
    user_project: UserProjectSchema = Depends(verify_admin_user_project_acess),
    photo: UploadFile = File(...),
):
foo()
....
```
with 
```python
class UserProjectSchema(BaseModel):
    user_id: int
    project_id: int
```

``` python
def verify_admin_user_project_acess(
    user_project: UserProjectSchema
) -> UserProjectSchema:
    if not is_admin(user_project.user_id, user_project.project_id):
        raise HTTPException(status_code=403, detail="User is not admin of this project")
    return user_project
  ```

When I test it, every way I try to put the data, I got 422.

Can someone explain me what is the shortest workaround please 🙏🏻 ?

**Reply by BloodyRain2k** [reply]
I think the issue with your code is that form types like `Form()` and `File()` cannot be mixed with normal types.
Try this and see if it works:
```
def add_project_user(
    status: str = Form(...),
    name: str = Form(...),
    email: str = Form(...),
    user_project: UserProjectSchema = Depends(verify_admin_user_project_acess),
    photo: UploadFile = File(...),
):
```

**Comment by MattdaVill** [comment]
Is there any chance that this will ever be solved or even acknowledged as an issue?

**Reply by TomFaulkner** [reply]
It's surprising to me that this hasn't been improved upon since the beginning.

I'm going to try to take https://github.com/fastapi/fastapi/discussions/8406#discussioncomment-10241628 and make a function that does it behind the scenes so there aren't duplicate or weird Pydantic schemas and I don't have to see the ugliness.

**Comment by TomFaulkner** [comment]
There was an earlier thread where what I think is a good solution was offered.

https://github.com/fastapi/fastapi/discussions/7268#discussioncomment-6993634

I shared it in a gist with a docstring, including an example, here.
https://gist.github.com/TomFaulkner/eeb8028d4afaca35dda85ebe5e7b76ef

Currently, it handles `Optional` in typehints, but not the new syntax. But, that should be easy enough to add.

**Reply by Brikas** [reply]
This works and it also reflects nicely in swagger! Why no one is noticing your comment from a year ago - there seemed to have been a lot of discussion recently on another thread?

**Comment by pythonweb2** [comment, marked as answer]
The simple answer to this question is that the feature was (mostly) implemented here: https://fastapi.tiangolo.com/tutorial/request-form-models/

However, it does not work with file uploads, which is something my original post did mention.

For example, if I try to do this:

```python
from typing import Annotated

import pydantic
import uvicorn
from fastapi import FastAPI, Form, UploadFile

app = FastAPI()


class FormData(pydantic.BaseModel):
    description: str
    tags: list[str]
    file: UploadFile


@app.post("/test_file_with_json")
async def test_form(data: Annotated[FormData, Form()]):
    print(data)
    return None


if __name__ == "__main__":
    uvicorn.run(app)
```

And then try to submit the form via SwaggerUI (the /docs URL), it won't work. This will work though (which feels a bit inconsistent, it should probably be documented somewhere).

```python
from typing import Annotated

import uvicorn
from fastapi import FastAPI, Form, UploadFile

app = FastAPI()


@app.post("/test_form")
async def test_form(
    description: Annotated[str, Form()],
    tags: Annotated[list[str], Form()],
    file: UploadFile,
):
    print(locals())
    return None


if __name__ == "__main__":
    uvicorn.run(app)
```

However, this also doesn't support sending an object in a form request, or a nested pydantic model. To do this, you can use the `pydantic.Json`.

```python
import pydantic
import uvicorn
from fastapi import FastAPI, Form, UploadFile

app = FastAPI()


class MetaData(pydantic.BaseModel):
    description: str
    tags: list[str]


# Example of an endpoint that uses the wrapper
@app.post("/test_file_with_json")
async def test_form(
    file: UploadFile,
    meta_data: Annotated[pydantic.Json[MetaData], Form()],
):
    print(file)
    print(meta_data)
    return {}


if __name__ == "__main__":
    uvicorn.run(app)
```

This works, if you know how to send the request, like this:

```python
import json
from pathlib import Path

import requests

if __name__ == "__main__":
    url = "http://localhost:8000/test_file_with_json"

    payload = json.dumps({"description": "string", "tags": ["string"]})
    with Path("scratch.txt").open("rb") as f:
        response = requests.post(url, headers={}, data={"meta_data": payload}, files={"file": f})

    print(response.text)
```

_But_, it will not work in SwaggerUI, since the schema that is generated from pydantic is interpreted by SwaggerUI to mean that a file upload is required, when really it should be a string:

<img width="1428" height="460" alt="image" src="https://github.com/user-attachments/assets/55eab24e-aee7-4015-86c8-528f1e536c74" />

The way to fix this is to make Pydantic generate the schema of the object, instead of the Json type wrapper schema. This is not super straightforward, but can be accomplished with a custom annotation:

```python
from typing import Annotated

import pydantic
from fastapi import FastAPI, Form, UploadFile
from pydantic import GetJsonSchemaHandler
from pydantic.json_schema import JsonSchemaValue
from pydantic_core import CoreSchema

app = FastAPI()


class NotAsJson:
    @classmethod
    def __get_pydantic_json_schema__(
        cls, core_schema: CoreSchema, handler: GetJsonSchemaHandler
    ) -> JsonSchemaValue:
        json_schema = handler(core_schema)
        return json_schema["contentSchema"] if "contentSchema" in json_schema else json_schema


class MetaData(pydantic.BaseModel):
    description: str
    tags: list[str]


@app.post("/test_file_with_json")
async def test_form(
    file: UploadFile,
    meta_data: Annotated[pydantic.Json[MetaData], NotAsJson, Form()],
):
    print(file)
    print(meta_data)
    return {}
```

Then we get the correct schema in the SwaggerUI page, and it will still work:

<img width="1429" height="637" alt="image" src="https://github.com/user-attachments/assets/4dbf25f8-c818-4695-a8c0-d546fe6e5d7e" />

---

I feel as though the documentation should be clearer about some of these topics, and maybe better support for JSON objects in multi-part form data should be added. I could probably update the docs if needed (tagging experts) @tiangolo @alejsdev @svlandeg @Kludex)

**Reply by pythonweb2** [reply]
Also tagging @YuriiMotov

**Reply by MatthewGiddes** [reply]
> The simple answer to this question is that the feature was (mostly) implemented here: https://fastapi.tiangolo.com/tutorial/request-form-models/
> 
> However, it does not work with file uploads, which is something my original post did mention.
> 
> For example, if I try to do this:
> 
> ```python
> from typing import Annotated
> 
> import pydantic
> import uvicorn
> from fastapi import FastAPI, Form, UploadFile
> 
> app = FastAPI()
> 
> 
> class FormData(pydantic.BaseModel):
>     description: str
>     tags: list[str]
>     file: UploadFile
> 
> 
> @app.post("/test_file_with_json")
> async def test_form(data: Annotated[FormData, Form()]):
>     print(data)
>     return None
> 
> 
> if __name__ == "__main__":
>     uvicorn.run(app)
> ```
> 
> And then try to submit the form via SwaggerUI (the /docs URL), it won't work. 

@pythonweb2 
When trying this example with curl everything looks fine to me using fastapi 0.128.0
`curl 0.0.0.0:8001/test_file_with_json -F description=des -F file=@test.txt -F tags=tag1`
Couldn't this just be a problem of how /docs assembles the request?

**Comment by ntairov** [comment]
for file uploads this approach works fine for me: 

```python
class JsonOrDictMixin:
    """Mixin that allows Form models to accept JSON string or dict input.

    This handles the Swagger UI limitation where Form models are sent as JSON strings
    instead of individual form fields when combined with file uploads.
    """

    @model_validator(mode="before")
    @classmethod
    def parse_json_string(cls, value: Any) -> Any:
        """Parse JSON string input into dict for Pydantic validation."""
        if isinstance(value, str):
            try:
                return json.loads(value)
            except json.JSONDecodeError as e:
                raise ValueError(f"Invalid JSON in form data: {e}") from e
        return value

```
```python
class MetaData(BaseModel, JsonOrDictMixin):
    description: str
    tags: list[str]
```

```python
@app.post("/test_file_with_json")
async def test_form(
    data: Annotated[schemas.MetaData, Form()],
    file: UploadFile,
):
    print(file)
    print(data)
    return {}
```

**Reply by brainbytes42** [reply]
Thank you @ntairov for your mixin, you saved my day!

It would be really great @tiangolo , if this could be integrated into FastApi until swaggers limitation is fixed. FastAPI is such a great and intuitive tool, so that this not working out of the box doesn't match the rest of the library. :-)
